### PR TITLE
[AKS] Add options for users to specify guardrails

### PR DESCRIPTION
**Pull request by NickKeller**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### Related command
`az aks create|update -g <RGNAME> -n <CLUSTERNAME> --guardrails-level Warning --guardrails-version v1.0.0 --guardrails-excluded-ns ns1,ns2 --enable-addons azure-policy`

### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
AKS

File: src/aks-preview/azext_aks_preview/_params.py [hunk 1]
```diff
@@ -410,13 +463,23 @@ def load_arguments(self, _):
...
-        c.argument('nodepool_asg_ids', validator=validate_application_security_groups, is_preview=True, help="application security groups for agentpool")
+        c.argument('nodepool_allowed_host_ports', validator=validate_allowed_host_ports,
+                   is_preview=True, help="allowed host ports for agentpool")
+        c.argument('nodepool_asg_ids', validator=validate_application_security_groups,
+                   is_preview=True, help="application security groups for agentpool")
         c.argument('node_public_ip_tags', arg_type=tags_type, validator=validate_node_public_ip_tags,
                    help='space-separated tags: key[=value] [key[=value] ...].')
+        c.argument('guardrails_level', arg_type=get_enum_type(guardrails_levels),
+                   help='The guardrails level, one of ["Off", "Warning", "Enforcement"]', is_preview=True)
+        c.argument('guardrails_version', type=str,
```
Comment (FumingZhang) [inline src/aks-preview/azext_aks_preview/_params.py:478]: Could omit this `type=str` as it's the default type.

Besides, may remove the help message here as you've written detailed help message in `_help.py`, and I think one of them will be overwritten.

File: src/aks-preview/azext_aks_preview/_params.py [hunk 2]
```diff
@@ -258,56 +274,76 @@ def load_arguments(self, _):
         c.argument('kubernetes_version',
                    completer=get_k8s_versions_completion_list)
         c.argument('dns_name_prefix', options_list=['--dns-name-prefix', '-p'])
-        c.argument('node_osdisk_diskencryptionset_id', options_list=['--node-osdisk-diskencryptionset-id', '-d'])
+        c.argument('node_osdisk_diskencryptionset_id', options_list=[
```
Comment (FumingZhang) [inline src/aks-preview/azext_aks_preview/_params.py:277]: Would be great if you could undo the format change (lines of less than 120 characters are allowed)

File: src/aks-preview/azext_aks_preview/managed_cluster_decorator.py [hunk 3]
```diff
@@ -2654,6 +2696,21 @@ def set_up_auto_upgrade_profile(self, mc: ManagedCluster) -> ManagedCluster:
...
+        level = self.context.get_guardrails_level()
+        # provided any value?
+        if (level is not None or version is not None or excludedNamespaces is not None) and mc.guardrails_profile is None:
+            mc.guardrails_profile = self.models.GuardrailsProfile(
+                level=level,
+                version=version
+            )
+        # replace values with provided values
+        if excludedNamespaces is not None:
+            mc.guardrails_profile.excluded_namespaces = excludedNamespaces.split(",")
```
Comment (FumingZhang) [inline src/aks-preview/azext_aks_preview/managed_cluster_decorator.py:2711]: Will the server side handle extra space after split properly?

File: src/aks-preview/azext_aks_preview/_help.py [hunk 4]
```diff
@@ -1000,6 +1013,15 @@
...
           long-summary: These certificates are used by Custom CA Trust features and will be added to trust stores of nodes. Requires Custom CA Trust to be enabled on the node.
+        - name: --guardrails-level
+          type: string
+          short-summary: The Guardrails Level. Accepted Values are [Off, Warning]. Requires azure policy addon to be enabled
+        - name: --guardrails-version
+          type: string
+          short-summary: The version of Guardrails to use. Default "v1.0.0" Use the ListGuardrailsVersions API to discover available versions
+        - name: --guardrails-excluded-ns
+          type: string
+          short-summary: Comma-separated list of Kubernetes namespaces to exclude from Guardrails. Use "[]" to clear a previously non-empty list
```
Comment (FumingZhang) [inline src/aks-preview/azext_aks_preview/_help.py:1024]: Some other options (like --cluster-autoscaler-profile, --api-server-authorized-ip-ranges, -nodepool-labels) use an empty string “” to reset the value. Is using "[]" the new norm?

**Review by FumingZhang — COMMENTED** [review]
Queued [live test pipeline](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=73461404&view=results) to verify the newly added test cases.

File: src/aks-preview/azext_aks_preview/managed_cluster_decorator.py (hunk 3 above)
Comment (NickKeller) [reply src/aks-preview/azext_aks_preview/managed_cluster_decorator.py:2711]: Changed to remove any extra spaces

File: src/aks-preview/azext_aks_preview/_help.py (hunk 4 above)
Comment (NickKeller) [reply src/aks-preview/azext_aks_preview/_help.py:1024]: Updated

File: src/aks-preview/azext_aks_preview/_params.py (hunk 1 above)
Comment (NickKeller) [reply src/aks-preview/azext_aks_preview/_params.py:433]: removed

File: src/aks-preview/azext_aks_preview/managed_cluster_decorator.py (hunk 3 above)
Comment (NickKeller) [reply src/aks-preview/azext_aks_preview/managed_cluster_decorator.py]: use extract_comma_separated_string instead

File: src/aks-preview/azext_aks_preview/managed_cluster_decorator.py [hunk 5]
```diff
@@ -3436,6 +3463,35 @@ def update_auto_upgrade_profile(self, mc: ManagedCluster) -> ManagedCluster:
             mc.auto_upgrade_profile.node_os_upgrade_channel = node_os_upgrade_channel
         return mc
 
+    def update_guardrails_profile(self, mc: ManagedCluster) -> ManagedCluster:
```
Comment (andyliuliming) [inline src/aks-preview/azext_aks_preview/managed_cluster_decorator.py:3466]: looks like there're many dup logic in update_guardrails_profile & set_up_guardrails_profile
possible to create a common function for those 2?

Comment (NickKeller) [reply src/aks-preview/azext_aks_preview/managed_cluster_decorator.py:3459]: refactored common code into helper funcion

**Review by andyliuliming — APPROVED** [review]

File: src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py [hunk 6]
```diff
@@ -1388,6 +1388,54 @@ def test_aks_nodepool_abort(self, resource_group, resource_group_location):
             self.check('powerState.code', 'Running')
         ])
 
+    @live_only()
```
Comment (zhoxing-ms) [inline src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:1391]: May I ask why do you mark these tests as `@live_only()`?

**Comment by zhoxing-ms** [comment]
@NickKeller Could you please resolve these CI issues?

File: src/aks-preview/HISTORY.rst [hunk 7]
```diff
@@ -22,6 +22,11 @@ Pending
 * fix: don't use current kube_proxy_config on UPDATE
 * GA update for Azure Monitor Metrics Addon (managed prometheus metrics) for AKS
 
+0.5.140
```
Comment (FumingZhang) [inline src/aks-preview/HISTORY.rst:25]: Please move this up a bit

File: src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py (hunk 6 above)
Comment (FumingZhang) [reply src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:1391]: Probably could remove this decorator (`@live_only()`), as you've already committed the recording files.

**Review by FumingZhang — APPROVED** [review]
LGTM

File: src/aks-preview/HISTORY.rst (hunk 7 above)
Comment (NickKeller) [reply src/aks-preview/HISTORY.rst]: moved

File: src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py (hunk 6 above)
Comment (NickKeller) [reply src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py]: Removed

**Comment by NickKeller** [comment]
> @NickKeller Could you please resolve these CI issues?

@zhoxing-ms These tests don't fail when I run `azdev test --no-exitfirst --discover aks-preview` locally. Is there another flag I should add?

**Comment by zhoxing-ms** [comment]
@NickKeller Could you please try to re-record the test `test_set_up_addon_profiles` in live mode? Such as `azdev test test_set_up_addon_profiles --live`

**Comment by NickKeller** [comment]
@zhoxing-ms I'm running it now, but I fail to see how that will help, considering it's a unit test, code of which I did not touch, is failing

**Comment by zhoxing-ms** [comment]
>         self.assertEqual(dec_mc_2, ground_truth_mc_2)
>         AssertionError: <azex[67 chars]iew.models._models_py3.ManagedCluster object at 0x7f9f279a0fa0> != <azex[67 chars]iew.models._models_py3.ManagedCluster object at 0x7f9f279a3190>
   src/aks-preview/azext_aks_preview/tests/latest/test_managed_cluster_decorator.py:3760: AssertionError
> - generated xml file: /home/cloudtest/.azdev/env_config/mnt/vss/_work/1/s/env/test_results.xml -
=========================== short test summary info ============================
> FAILED src/aks-preview/azext_aks_preview/tests/latest/test_managed_cluster_decorator.py::AKSPreviewManagedClusterCreateDecoratorTestCase::test_set_up_addon_profiles

This seems to be some logic issue from the unit testing itself  + @FumingZhang help take a look~

**Comment by FumingZhang** [comment]
Checked the error, should change `"True"` to `"true"` in line 3740 of file `test_managed_cluster_decorator.py`. Probably caused by change in previous PR #6311.

**Comment by FumingZhang** [comment]
Replaced by PR #6357

**Pull request closed without merging**